### ISSUE TITLE
Remove TextMate.selected_paths_array from shelltokenize.rb, but maybe keep functionality?

### ISSUE THREAD

**Issue by noniq** [textmate/bundle-support.tmbundle#25]
[This](https://github.com/textmate/bundle-support.tmbundle/blob/f656497e00b07d6ca5c4d6b9b9ee49e8e985f0da/Support/shared/lib/shelltokenize.rb#L93-L111) is more or less a duplicate of `TextMate.selected_files` from `textmate.rb`. It is currently only used in a few places in the CVS, Mercurial, SVK and SVN bundles (sometimes via `TextMate.selected_paths_for_shell`).

However, there is one difference: If no files are selected, `selected_path_array` returns an array with a single entry for `ENV["TM_FILEPATH"]`, or an empty array if this is not set. `selected_files` returns `nil` in those cases.

I wonder whether this functionality (“Give me an array of selected files, falling back to the current file if nothing is selected“) is so common that it should be part of the “official” bundle support API?

For example, the Git bundle has its own rather elaborate implementation that even supports different fallbacks (current file vs. project directory) and optional unique-filtering: https://github.com/textmate/git.tmbundle/blob/d1db42c2d71948662098183a6df519fb53a7a15b/Support/lib/git.rb#L114-L137

What do you think?

**Comment by sorbits** [comment]
I assume the library would be something like this:

```ruby
require 'shellwords'

module TextMate
  def TextMate.file_selection
    ENV.has_key?('TM_SELECTED_FILES') ? Shellwords.shellwords(ENV['TM_SELECTED_FILES']) : nil
  end

  def TextMate.file_selection_or_current
    return Shellwords.shellwords(ENV['TM_SELECTED_FILES']) if ENV.has_key?('TM_SELECTED_FILES')
    return [ ENV['TM_FILEPATH'] ] if ENV.has_key?('TM_FILEPATH')
    nil
  end

  def TextMate.file_selection_or_project
    return Shellwords.shellwords(ENV['TM_SELECTED_FILES']) if ENV.has_key?('TM_SELECTED_FILES')
    return [ ENV['TM_PROJECT_DIRECTORY'] ] if ENV.has_key?('TM_PROJECT_DIRECTORY')
    return [ ENV['TM_DIRECTORY'] ] if ENV.has_key?('TM_DIRECTORY')
    nil
  end
end
```

I could see a minor advantage of having commands use such library, even though inlining the relevant code is probably not that much more, but probably a bit more arcane.

I’m btw definitely in favor of returning `nil` over empty array, as the former can more easily be tested against.

**Comment by noniq** [comment]
Exactly. Btw, there is already `TextMate.selected_files` in textmate.rb: https://github.com/textmate/bundle-support.tmbundle/blob/f656497e00b07d6ca5c4d6b9b9ee49e8e985f0da/Support/shared/lib/textmate.rb#L213-L217. Maybe  simply add an optional `fallback:` parameter so that it can be called like this:

```ruby
TextMate.selected_files # default (nil if no files selected)
TextMate.selected_files(fallback: :current_file) 
TextMate.selected_files(fallback: :project_dir) 
```

On the other hand, having explicit methods would be more explicit. So not sure what’s the best way.

```

**Comment by sorbits** [comment]
I like having different method names for the different options, 
especially with ruby where options are untyped (so one can pass in 
virtually anything w/o getting errors).

Encoding the options in the method names also makes it easier to search 
for the code which uses the options.

And given that we already have TextMate.selected_files, let’s 
definitely move everything to moving that, instead of this shelltokenize 
stuff.

The uniquing done by the Git bundle, maybe this was for 1.x projects 
where the user could add the same file twice, but there should be no 
reason to check for duplicates in TM_SELECTED_FILES with 2.0.
